File: app/utils/csp.py
# CSP自動計算ロジック（要件定義書 6章）

def calc_city_csp(rank: int, participants: int) -> int:
    if participants >= 1 and rank == 1:
        csp = 100
    elif participants >= 4 and rank == 2:
        csp = 75
    elif participants >= 9 and rank <= 4:
        csp = 50
    elif participants >= 17 and rank <= 8:
        csp = 25
    elif participants >= 33 and rank <= 16:
        csp = 15
    else:
        csp = 0
    return csp


TRAINER_ANNUAL_LIMIT = 30


def calc_trainer_csp(rank: int, participants: int, current_total: int) -> int:
    csp = 0
    if participants >= 2 and rank == 1:
        csp = 15
    elif participants >= 3 and rank == 2:
        csp = 12
    elif participants >= 8 and rank <= 4:
        csp = 10
    elif participants >= 13 and rank <= 8:
        csp = 8
    elif participants >= 26 and rank <= 16:
        csp = 6
    elif participants >= 41 and rank <= 32:
        csp = 5
    elif participants >= 56 and rank <= 48:
        csp = 4
    remaining = TRAINER_ANNUAL_LIMIT - current_total
    return max(0, min(csp, remaining))

File: js/csp.js
// ============================================================
// csp.js — クライアント側 CSP 計算（プレビュー用）
// ============================================================
// GAS の CSP.gs と同一アルゴリズム（要件定義書 6-2, 6-3 準拠）

const CITY_CSP_TABLE = [
  [1,  1, 100],
  [2,  4,  75],
  [4,  9,  50],
  [8, 17,  25],
  [16,33,  15],
];

const TRAINER_CSP_TABLE = [
  [1,  2, 15],
  [2,  3, 12],
  [4,  8, 10],
  [8, 13,  8],
  [16,26,  6],
  [32,41,  5],
  [48,56,  4],
];

const TRAINER_ANNUAL_LIMIT = 30;

/**
 * シティリーグ CSP 計算（プレビュー用）
 * @param {number} rank
 * @param {number} participants
 * @returns {number}
 */
function previewCityCSP(rank, participants) {
  const r = parseInt(rank);
  const p = parseInt(participants);
  if (!r || !p || r < 1 || p < 1) return 0;
  for (const [rankLimit, pMin, csp] of CITY_CSP_TABLE) {
    if (p >= pMin) {
      if (r <= rankLimit) return csp;
    } else {
      break;
    }
  }
  return 0;
}

/**
 * トレーナーズリーグ CSP 計算（プレビュー用）
 * @param {number} rank
 * @param {number} participants
 * @param {number} currentTotal - 現在の年間累計（この試合を除く）
 * @returns {number}
 */
function previewTrainerCSP(rank, participants, currentTotal = 0) {
  const r = parseInt(rank);
  const p = parseInt(participants);
  if (!r || !p || r < 1 || p < 2) return 0;
  let cspEarned = 0;
  for (const [rankLimit, pMin, csp] of TRAINER_CSP_TABLE) {
    if (p >= pMin) {
      if (r <= rankLimit) { cspEarned = csp; break; }
    } else {
      break;
    }
  }
  const remaining = TRAINER_ANNUAL_LIMIT - currentTotal;
  return Math.max(0, Math.min(cspEarned, remaining));
}

/**
 * 日時をローカライズ表示
 * @param {string} iso
 * @returns {string}
 */
function formatDate(iso) {
  if (!iso) return '';
  try {
    return new Date(iso).toLocaleDateString('ja-JP', {
      year: 'numeric', month: 'numeric', day: 'numeric',
      hour: '2-digit', minute: '2-digit',
    });
  } catch {
    return iso;
  }
}
